### PR TITLE
[CUDA] Support CUDA 9 and test different CUDA versions at CI

### PR DESCRIPTION
**Pull request by StrikerRUS**
I think testing different CUDA versions will greatly help us to be confident in CUDA code.

Linking #3431.

CUDA < 10.0 fails to recognize `7.5` architecture.
```
[  7%] Building CUDA object CMakeFiles/histo_16_64_256-fulldata_sp_const.dir/src/treelearner/kernels/histogram_16_64_256.cu.o
nvcc fatal   : Unsupported gpu architecture 'compute_75'
```
Refer to https://github.com/Kitware/CMake/blob/36bb0e32d7e3976eed424078cf5cac459651f0f1/Modules/FindCUDA/select_compute_arch.cmake#L50-L60.

### CODE REVIEW

File: .github/workflows/cuda.yml [hunk 1]
```diff
@@ -26,10 +26,13 @@ jobs:
         include:
           - method: source
             python_version: 3.6
+            cuda_version: "11.2.0"
           - method: pip
             python_version: 3.7
+            cuda_version: "10.0"
```
Comment (StrikerRUS) [inline .github/workflows/cuda.yml:32]: GitHub Actions treats values written as `x.0` like `int`s, so we need explicitly set them to strings.

File: CMakeLists.txt [hunk 2]
```diff
@@ -156,7 +156,14 @@ if(USE_CUDA)
...
     LIST(APPEND CMAKE_CUDA_FLAGS -Xcompiler=${OpenMP_CXX_FLAGS} -Xcompiler=-fPIC -Xcompiler=-Wall)
-    CUDA_SELECT_NVCC_ARCH_FLAGS(CUDA_ARCH_FLAGS 6.0 6.1 6.2 7.0 7.5+PTX)
+
+    set(CUDA_ARCHS "6.0" "6.1" "6.2" "7.0")
+    if(CUDA_VERSION VERSION_GREATER_EQUAL "10.0")
+        list(APPEND CUDA_ARCHS "7.5")
+    endif()
+    list(POP_BACK CUDA_ARCHS CUDA_LAST_SUPORTED_VERSION)
+    list(APPEND CUDA_ARCHS "${CUDA_LAST_SUPORTED_VERSION}+PTX")
+    CUDA_SELECT_NVCC_ARCH_FLAGS(CUDA_ARCH_FLAGS ${CUDA_ARCHS})
```
Comment (StrikerRUS) [inline CMakeLists.txt:166]: cc @ChipKerchner 
I hope this change will not break anything on POWER machines.

Comment (ChipKerchner) [reply CMakeLists.txt:166]: I don't think so

File: .ci/setup.sh [hunk 3]
```diff
@@ -80,9 +80,15 @@ else  # Linux
...
-            cmake \
             curl \
+            lsb-release \
+            software-properties-common \
             wget
+        curl -sL https://apt.kitware.com/keys/kitware-archive-latest.asc | apt-key add -
+        apt-add-repository "deb https://apt.kitware.com/ubuntu/ $(lsb_release -cs) main" -y
+        apt-get update
+        apt-get install --no-install-recommends -y \
+            cmake
```
[the remark this answers is not in the document]
Comment (jameslamb) [reply .ci/setup.sh:91]: blegh ok, that's alright

File: .github/workflows/cuda.yml (hunk 1 above)
Comment (jameslamb) [reply .github/workflows/cuda.yml:32]: :scream: that's horrible haha

**Review by jameslamb — APPROVED** [review]
good idea, thanks

File: CMakeLists.txt (hunk 2 above)
Comment (StrikerRUS) [reply CMakeLists.txt:166]: @ChipKerchner I'm so sorry for my poor English skills, but could you please rephrase? Do you think that
(1) this change can potentially hurt compilation on POWER systems
OR
(2) this change should be OK for POWER systems
?

And if (1), how this change can be harmful to POWER machines?

Comment (ChipKerchner) [reply CMakeLists.txt:166]: I think this change is OK for POWER systems.

Comment (StrikerRUS) [reply CMakeLists.txt:166]: OK, got it! Thanks a lot for finding time to comment!

**Pull request merged**